core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+package zio.internal
+
+import zio.internal.ConcurrentWeakHashSet.Ref
+
+import java.lang.ref.WeakReference
+import scala.collection.mutable
+import java.util.concurrent.locks.ReentrantLock
+import java.lang.ref.ReferenceQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+private[zio] object ConcurrentWeakHashSet {
+
+  private[internal] val MaxConcurrencyLevel: Int     = 1 << 16
+  private[internal] val MaxSegmentSize: Int          = 1 << 30
+  private[internal] val DefaultInitialCapacity: Int  = 16
+  private[internal] val DefaultLoadFactor: Float     = 0.75f
+  private[internal] val DefaultConcurrencyLevel: Int = 16
+
+  /**
+   * Reference node that links elements in the set with the same hash code
+   *
+   * @param hash
+   *   hash code of the element
+   * @param element
+   *   value passed to the weak reference
+   * @param queue
+   *   reference queue that will be used to track garbage collected elements
+   * @param nextRef
+   *   next reference node in the chain
+   * @tparam V
+   *   type of the element
+   */
+  protected class Ref[V](
+    val hash: Int,
+    element: V,
+    queue: ReferenceQueue[V],
```

Comment:
It's fine to use a reference queue, but we never block via its `remove` methods.